Src/StartMenu/StartMenuDLL/JumpLists.cpp

```diff
@@ -617,9 +544,50 @@ bool ExecuteJumpItem( const CItemManager::ItemInfo *pAppInfo, const CJumpItem &i
 
 	if (item.type==CJumpItem::TYPE_LINK)
 	{
-		// invoke the link through its context menu
+		//  Name:     System.AppUserModel.HostEnvironment -- PKEY_AppUserModel_HostEnvironment
+		//  Type:     UInt32 -- VT_UI4
+		//  FormatID: {9F4C2855-9F79-4B39-A8D0-E1D42DE1D5F3}, 14
+		static const PROPERTYKEY PKEY_AppUserModel_HostEnvironment = { {0x9F4C2855, 0x9F79, 0x4B39, {0xA8, 0xD0, 0xE1, 0xD4, 0x2D, 0xE1, 0xD5, 0xF3}}, 14 };
+
+		//  Name:     System.AppUserModel.ActivationContext -- PKEY_AppUserModel_ActivationContext
+		//  Type:     String -- VT_LPWSTR
+		//  FormatID: {9F4C2855-9F79-4B39-A8D0-E1D42DE1D5F3}, 20
+		static const PROPERTYKEY PKEY_AppUserModel_ActivationContext = { {0x9F4C2855, 0x9F79, 0x4B39, {0xA8, 0xD0, 0xE1, 0xD4, 0x2D, 0xE1, 0xD5, 0xF3}}, 20 };
+
 		CComQIPtr<IContextMenu> pMenu(item.pItem);
-		if (!pMenu) return false;
+		CStringA params;
+
+		CComQIPtr<IShellLink> pLink(item.pItem);
+		if (pLink)
+		{
+			CComQIPtr<IPropertyStore> store(pLink);
+			if (store)
+			{
+				auto appId = GetPropertyStoreString(store, PKEY_AppUserModel_ID);
+				if (!appId.IsEmpty())
+				{
+					CComPtr<IShellItem2> target;
+					if (SUCCEEDED(SHCreateItemInKnownFolder(FOLDERID_AppsFolder, 0, appId, IID_PPV_ARGS(&target))))
+					{
+						ULONG modern = 0;
+						if (SUCCEEDED(target->GetUInt32(PKEY_AppUserModel_HostEnvironment, &modern)) && modern)
+						{
+							CComQIPtr<IContextMenu> targetMenu;
+							if (SUCCEEDED(target->BindToHandler(nullptr, BHID_SFUIObject, IID_PPV_ARGS(&targetMenu))))
+							{
+								pMenu = targetMenu;
+								params = CT2CA(GetPropertyStoreString(store, PKEY_AppUserModel_ActivationContext));
+							}
+						}
+					}
+				}
+			}
+		}
+
```

Comment:
No exceptions in case something goes horribly wrong? That seems like a lot of ifs without much handling if things turn out differently.
```suggestion
try {
CStringA params;
// ... 
catch(exception) {
 throw new Exception("uh oh");
 }
```